### DISCUSSION TITLE
Seeking advice on serializing game assets - An unusual use-case

### DISCUSSION THREAD

**Discussion by SirusDoma** [capnproto/capnproto#2141] [Q&A]
I’m currently in the middle of deciding whether to pick capnproto (or [bebop](https://github.com/betwixt-labs/bebop)) for my existing project, I did a little research but I still can’t get my answer based on my use case. Here’s my scenario:

I’m developing game using c++ and my game assets are serialized using json. The json files contains metadata of our game assets/resource files, for example a sprite metadata has information about transform position, scale, rotation, as well as texture uv. At the simplest form, it look like this:

```json
{
  "type": "Sprite",
  "require": {
    "texture": "Path/to/some/texture.bmp"
  },
  "attributes": {
    "transform": {
      "position": {
        "x": 0,
        "y": 0
      }
    },
    "texCoords": {
      "x": 0,
      "y": 0,
      "width": 800,
      "height": 600
    }
  }
}
```

This works, but it the size (and performance impact) is not ideal as our json files grow in numbers. So I’m looking serialization based on these few things:

- Smallest footprint of the serialized binary size in disk (and in memory usage) is the most important
- De-serialization performance
- Ability to serialize and deserialize the models as json files for easy access to editing/authoring assets purposes and convert them into binary later where the game can consume them
- The structure of the resource metadata will evolve in the future game version. I want to maintain compatibility in such scenario where newer assets being consumed by older game client.
- I use c++17 so we can’t have generated code to anything newer than that

Is capnproto suitable for this? And what's your take of capnproto compared to other competitors, particularly, protobuf and bebop for this scenario?

**Comment by alexshpilkin** [comment]
The use case is probably not as unusual as you think: Google’s game technology lab made a competing zero-copy serialization format, FlatBuffers, for exactly that use case. You may be interested in their discussion of the differences: https://github.com/google/flatbuffers/issues/2; but keep an eye on Kenton’s comparison too: https://capnproto.org/news/2014-06-17-capnproto-flatbuffers-sbe.html.

**Reply by SirusDoma** [reply]
This is interesting, `FlatBuffer` is actually being used by cocos2d-x too and make me certain using one of these library is a right approach to my serialization problem.

Both comparison you mentioned are made in 2014, I'm wondering if these information still relevant. On the other hand, I found [bebop](https://github.com/betwixt-labs/bebop), which compare itself against capnproto and FlatBuffers like this:

> We don’t benchmark against FlatBuffers or Cap'n Proto for a few reasons:
> 
> - They are impossible to benchmark against each other without making an assumption about how often you want to access the data, and which parts of it you want to access. But this means Bebop, Cap'n Proto, FlatBuffers can exist side-by-side / solve different problems: what FlatBuffers and Cap'n Proto do makes sense if you want to access only parts of your objects in limited specific ways, what we do is better if you're always interested in the whole packet.
> 
> - We don’t benchmark against Capt’n Proto because it does not have a stable web-based implementation, at least not one that has the features that make it so fast natively, so there is nothing to compare.
> 
> Basically, Bebop is great for message oriented applications or where you need the entirety of your packet deserialized in a single step. So we can only benchmark against similar formats (JSON, Protobufs, MsgPack, etc.)

It can be found in [here](https://news.ycombinator.com/item?id=39678774) and [here](https://www.reddit.com/r/programming/comments/1bcvrnp/comment/kuiirit/). My main take away is the following part:

> what FlatBuffers and Cap'n Proto do makes sense if you want to access only parts of your objects in limited specific ways, what we do is better if you're always interested in the whole packet

Which sounds more relatable to my use-case due to the need of the entirety of my asset message to parse the game asset as a whole because it definitely cannot be load by just accessing some part of it.  

I'm trying to reach the author of bebop to elaborate this further but I don't expect the answer anytime soon since their discord community is not really active and I would really love to what Kenton's say specifically about this.
